target/linux/realtek/files-5.10/drivers/net/phy/rtl83xx-phy.c

```diff
@@ -3165,13 +3294,295 @@ int rtl9300_sds_cmu_band_get(int sds)
+	{0, 0x1f, 0x05A9}, {0, 0x10, 0x7F04}, {0, 0x13, 0xA854}, {0, 0x14, 0x3202},
+	{0, 0x1f, 0x05AA}, {0, 0x10, 0x032F}, {0, 0x15, 0x7F52}, {0, 0x1f, 0x05AB},
+	{0, 0x16, 0x01E0}, {0, 0x1f, 0x05AC}, {0, 0x15, 0x4380}, {0, 0x1f, 0x05AD},
+	{0, 0x10, 0x4321}, {0, 0x11, 0x010F}, {0, 0x12, 0x88FF}, {0, 0x13, 0x4208},
+	{0, 0x1f, 0x05AF}, {0, 0x14, 0xF0F3}, {0, 0x15, 0xF2F0}, {0, 0x1f, 0x0504},
+	{0, 0x11, 0x000F}, {0, 0x1f, 0x0500}, {0, 0x10, 0x1400}, {0, 0x10, 0x1403},
+	{0, 0x1f, 0x0000}, {0, 0x1e, 0x0001},
+};
+
+/*
+ * RTL9300-specific patches for the RTL828D for XSGMII
+ */
+phy_config rtl8218d_a_xsgmii_rtl9300[] = {
+	{0, 0x1f, 0x0000}, {0, 0x1e, 0x0008}, {0, 0x1f, 0x0400}, {0, 0x12, 0x71D0},
+	{0, 0x1f, 0x0500}, {0 ,0x12, 0x71D0}, {0, 0x1f, 0x0401}, {0, 0x16, 0x3053},
+	{0, 0x1f, 0x0480}, {0, 0x13, 0x0200}, {0, 0x1f, 0x0485}, {0, 0x13, 0x0804},
+	{0, 0x1f, 0x04B8}, {0, 0x11, 0x2C27}, {0, 0x12, 0x0100}, {0, 0x13, 0xE015},
+	{0, 0x17, 0x000A}, {0, 0x1f, 0x04B9}, {0, 0x10, 0x7F04}, {0, 0x13, 0xA854},
+	{0, 0x14, 0x3A02}, {0, 0x1f, 0x04BA}, {0, 0x10, 0x032F}, {0, 0x11, 0x0121},
+	{0, 0x15, 0x7E12}, {0, 0x17, 0x808F},
+	{0, 0x1f, 0x04BB}, {0, 0x16, 0x01E0}, {0, 0x1f, 0x04BC}, {0, 0x10, 0x02C4},
+	{0, 0x1f, 0x04BD}, {0, 0x10, 0x4321}, {0, 0x11, 0x010F}, {0, 0x12, 0x88FF},
+	{0, 0x13, 0x4208}, {0, 0x1f, 0x04BF}, {0, 0x14, 0xF0F3}, {0, 0x15, 0xF2F0},
+	{0, 0x1f, 0x0404}, {0, 0x11, 0x000F}, {0, 0x1f, 0x0486}, {0, 0x10, 0x001F},
+	{0, 0x1f, 0x0400}, {0, 0x10, 0x1700}, {0, 0x10, 0x1703}, {0, 0x1f, 0x0000},
+	{0, 0x1e, 0x0000},
+};
+
+static void rtl9300_phy_patch(struct phy_device *phydev, phy_config c[], int entries)
+{
+	int i;
+
+	for (i = 0; i < entries; ++i)
+		phy_write_paged(phydev + c[i].phy, REALTEK_PATCH_PAGE, c[i].reg, c[i].data);
+}
+
+static int rtl9300_rtl821d_phy_setup(struct phy_device *phydev, int phy_mode)
+{
+	u32 model;
+
+	rtl8218d_serdes_mode_get(phydev);
+
+	model = rtl8218d_serdes_model_get(phydev);
+
+	if (model & BIT(7)) {  // Is RTL821D_NMP?
+		switch(phy_mode) {
+		case PHY_INTERFACE_MODE_QSGMII:
+			rtl9300_phy_patch(phydev, rtl8218d_bT_qsgmii_rtl9300,
+					  sizeof (rtl8218d_bT_qsgmii_rtl9300) / sizeof(phy_config));
+			break;
+		case PHY_INTERFACE_MODE_XGMII:
+			rtl9300_phy_patch(phydev, rtl8218d_bT_xsgmii_rtl9300,
+					  sizeof (rtl8218d_bT_xsgmii_rtl9300) / sizeof(phy_config));
+			break;
+		default:
+			pr_err("%s: Unsupported PHY mode\n", __func__);
+			return -EINVAL;
+		}
+	} else { // Normal RTL821D
+		switch(phy_mode) {
+		case PHY_INTERFACE_MODE_QSGMII:
+			rtl9300_phy_patch(phydev, rtl8218d_a_qsgmii_rtl9300,
+					  sizeof (rtl8218d_a_qsgmii_rtl9300) / sizeof(phy_config));
+			break;
+		case PHY_INTERFACE_MODE_XGMII:
+			rtl9300_phy_patch(phydev, rtl8218d_a_xsgmii_rtl9300,
+					  sizeof (rtl8218d_a_xsgmii_rtl9300) / sizeof(phy_config));
+			break;
+		default:
+			pr_err("%s Unsupported PHY mode\n", __func__);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+int rtl9300_configure_8218d(struct phy_device *phydev)
```

Comment:
don't forget to reference this function in line [4572](https://github.com/openwrt/openwrt/pull/10319/commits/aa0640117cd3ec79faf39485a8b2eb6ec45857ed#diff-f2e3b446d6bf87d07398d90e24463324c5e7a1a033610512f668d1046bfdab25R4572) which you accidentally did in improve rtl8218d patch; i think this function is unused in this patch